assembly/src/assembler/mod.rs

```diff
@@ -112,257 +313,460 @@ impl Assembler {
+    /// Like [Assembler::compile], but also takes an [AssemblyContext] to configure the assembler.
+    pub fn assemble_in_context(
+        &mut self,
+        source: impl Compile,
+        context: &mut AssemblyContext,
+    ) -> Result<Program, Report> {
+        let opts = CompileOpts {
+            warnings_as_errors: context.warnings_as_errors(),
+            ..CompileOpts::default()
+        };
+        self.assemble_with_opts_in_context(source, opts, context)
+    }
+
+    /// Compiles the provided module into a [Program] using the provided options.
+    ///
+    /// The resulting program can be executed on Miden VM.
     ///
     /// # Errors
-    /// Returns an error if the compilation of the specified program fails.
-    #[instrument("compile_ast", skip_all)]
-    pub fn compile_ast(&self, program: &ProgramAst) -> Result<Program, AssemblyError> {
-        // compile the program
-        let mut context = AssemblyContext::for_program(Some(program));
-        let program_root = self.compile_in_context(program, &mut context)?;
+    ///
+    /// Returns an error if parsing or compilation of the specified program fails, or the options
+    /// are invalid.
+    pub fn assemble_with_opts(
+        &mut self,
+        source: impl Compile,
+        options: CompileOpts,
+    ) -> Result<Program, Report> {
+        let mut context = AssemblyContext::default();
+        context.set_warnings_as_errors(options.warnings_as_errors);
+
+        self.assemble_with_opts_in_context(source, options, &mut context)
+    }
 
-        // convert the context into a call block table for the program
-        let cb_table = context.into_cb_table(&self.proc_cache.borrow())?;
+    /// Like [Assembler::compile_with_opts], but additionally uses the provided [AssemblyContext]
+    /// to configure the assembler.
+    #[instrument("assemble_with_opts_in_context", skip_all)]
+    pub fn assemble_with_opts_in_context(
+        &mut self,
+        source: impl Compile,
+        options: CompileOpts,
+        context: &mut AssemblyContext,
+    ) -> Result<Program, Report> {
+        if options.kind != ModuleKind::Executable {
+            return Err(Report::msg(
+                "invalid compile options: compile_with_opts requires that the kind be 'executable'",
+            ));
+        }
+
+        let program = source.compile_with_opts(CompileOpts {
+            // Override the module name so that we always compile the executable
+            // module as #exec
+            path: Some(LibraryPath::from(LibraryNamespace::Exec)),
+            ..options
+        })?;
+        assert!(program.is_executable());
+
+        // Remove any previously compiled executable module and clean up graph
+        let prev_program = self.module_graph.find_module_index(program.path());
+        if let Some(module_index) = prev_program {
+            self.module_graph.remove_module(module_index);
+            self.procedure_cache.remove_module(module_index);
```

Comment:
A couple things:

1. It seems to me that we are really abusing the notion of what an "executable" module is with the approach to compilation described above. Note scripts in a transaction are not really executables, they are conceptually libraries with a well-defined/conventional entrypoint (we are just abusing the "main" procedure of executable modules for this purpose, but it could just as easily be the case that we expect note script modules to export a procedure called `main` with zero arguments). The only "real" executable module here is the transaction script (and as we've discussed elsewhere, it might even be the case that the only executable is the kernel itself, but I don't want to muddy things too much here). To provide some intuition for what I mean: consider how executables are compiled/used for other platforms - it is not possible/supported to compile an executable that contains other executables (not in the sense that you compiled executable `A`, then compiled executable `B` by pointing to `A` and telling the compiler to link it in to `B`, instead `A` would be compiled as a library, and linked in to `B` like any other library, with `B` acting as a driver for `A` - i.e. this is what is commonly referred to as a "multi-call binary").
2. Prior to this PR, executable modules didn't have a unique name, i.e. it was always `#exec` or whatever. This naturally causes conflicts if trying to add multiple such modules to the same context. We can easily support multiple executable modules together (to the extent that even makes sense), simply by requiring that they be uniquely named (in which case the code referenced here could be removed). This code basically exists to account for the way executable modules were previously being used. There are other ways we could approach it, but the assembler is sort of structured to compile a single executable or library at a time, but allow for reusing the assembler instance to compile multiple executables/libraries consecutively and reuse shared compilation artifacts. There are other benefits to isolating assembler contexts this way (better performance, smaller footprint, etc.), but it isn't an inherent limitation, just one that fell out naturally from looking at how executable vs kernel vs library modules are structured/used - the semantics of executables and kernels sort of implicitly assume that they are the only module of their kind in a single context. It's not that we can't support a different set of semantics, just that the scenario you are describing runs a bit counter to how Miden Assembly documents those type of modules, and as a result, how the assembler works in this PR.

If the assumption is that we have to support multiple executables in the same context, I can make some modifications to allow for that (again, based on the assumption that the modules have unique names, perhaps generated, or derived from their path, or whatever). I think my recommendation would instead be to move to a model where note scripts are instead structured/compiled as library modules with a single procedure exported by convention, e.g. `main`, that acts as the entrypoint, but I can see why it felt natural to use `begin..end` for that. Long story short, we (as discussed elsewhere) almost certainly need to revisit the different module types, how they are meant/allowed to be used, etc., or this will continue to be a pain point.

Reply:
> One thing I did notice is that we don't seem to have Assembler::build_cb_table() method any more (at least not one exposed publicly). We should probably add it.

It is no longer needed, at least based on all usages in this repo:

* The `assemble_module` API allows you to compile a module and get back the MAST roots of all its exports
* The `assemble_*` APIs produce a `Program` that contains the `CodeBlockTable` for that program
* The `procedure_cache` API lets you directly query the procedure cache for the MAST roots of any procedure that has been compiled so far (or iterate over all contents of the cache). This can be used to construct a `CodeBlockTable` as well, though care must be taken when doing so, so as to only include code you actually need in the resulting artifact.

If there are other use cases we need to support, it would be good to document what those are here in the assembler code, so that we can take those into account both in the API itself, as well as in any future PR reviews which modify the assembler and its internals.

Reply:
> The only "real" executable module here is the transaction script (and as we've discussed elsewhere, it might even be the case that the only executable is the kernel itself, but I don't want to muddy things too much here).

I actually agree that the only real executable here is the `main` function of the transaction kernel - everything else gets invoked form it.

> It seems to me that we are really abusing the notion of what an "executable" module is with the approach to compilation described above

The distinction I wanted to make originally (and maybe choosing the word "executable" wasn't right for this) was that there are modules which get reduced to a single MAST root and there are modules which get reduced to multiple MAST roots. I called the former "executable" and the latter "library" and notes fell into the former category (i.e., a note script must reduce to a single MAST root).

Another point of difference is that "executable" modules is not something that could be imported by other modules (i.e., the way things currently work, it doesn't make sene to import a note script into another program). So, to summarize, the rules I used:

- An "executable" module is a module with code which reduces to a single MAST root and cannot be imported into other modules.
- A "library" modules is a module which reduces to one or more MAST roots and can be imported into other modules.

> I think my recommendation would instead be to move to a model where note scripts are instead structured/compiled as library modules with a single procedure exported by convention, e.g. `main`, that acts as the entrypoint, but I can see why it felt natural to use `begin..end` for that.

I think we we can do this - the question is do we still want to enforce some of the above rules. Specifically:
1. A note module must have a `main` procedure.
2. If there is a `main` procedure, no other procedure should be exported from the module.

The second rule is probably not a hard rule as exporting extra stuff is technically not an error (but it may be good to show it as a warning).

Then the question becomes if the assembler treats modules with `main` procedure differently from others - do they such modules deserve their own kind?

> If the assumption is that we have to support multiple executables in the same context, I can make some modifications to allow for that

I don't think that's needed as we are planning to move to MAST-based modules, and there this will no longer be relevant (i.e., each note script will be compiled into a separate component). But I do think that within the MAST format we should be able to define an entrypoint for note execution. This could be a part of the format itself (i.e., via `entrypoint` property or something else) or done by convention (i.e., with `main` procedure). I don't have a strong opinion formed on this yet - but it does seem like relying on convention is a bit more error-prone.

Reply:
I think it might help to break this down a little bit further. Essentially, we have two different sets of intersection issues here:

1. What types of modules/artifacts can be defined in Miden Assembly syntax, how they can be used together (i.e. semantics, restrictions on certain combinations), and how that translates into MAST/packaging/something executable by the VM. This is all fundamental to Miden Assembly, the instruction set of the VM, and how the VM expects to load and execute code compiled to MAST - completely isomorphic to the rollup or any other higher-level abstractions.
2. What types of artifacts/abstractions need to exist in order to support the rollup (or other similar efforts), how they can be used together, and how that translates into the primitives provided by MASM/MAST and the VM. Additionally, how this relates to the packaging format expected by the VM.

Looking solely at point 1 above, these are my observations:

* We need "bare metal" executables (i.e. kernels) which provide the following:
  - An entrypoint that the VM can invoke to perform cold-start initialization (e.g. rodata init, things like that)
  - An entrypoint that the VM can invoke to execute code (this could be the same entrypoint as above, or separate)
  - A set of procedures which provide the functionality available via `syscall`
* We need executables that target a specific kernel, and are designed to be loaded/executed by that kernel. Unlike the "bare-metal" executable of a kernel, these executables expect a certain amount of runtime support in the form of syscalls, and their lifecycle and behavior is dictated by the kernel. In both cases however, the executable entrypoint is defined using `begin..end`, and the module defining that entrypoint cannot export any procedures.
* Lastly, we need libraries which can be linked into both kinds of executable. Specific procedures in a library may be unsupported in certain contexts based on its capability set (as mentioned in another comment of mine, this refers to whether a procedure requires the use of an instruction that is restricted in certain contexts).
* Only one kernel can exist at a time

Aside from it being a bit unclear to me how best to describe the manner in which a kernel determines what it should do when its entrypoint is executed, I think it is fine to leave this unspecified for now. From the POV of the VM, it is instantiated with an executable (from the VM perspective, it probably doesn't even need to understand the distinction between a kernel and non-kernel executable), a set of inputs, etc., and is told to execute it - everything after that is determined by the executable module itself. In this way, compiling a simple MASM executable and executing it via the VM "just works", because that executable is treated as a kernel. More complex programs with an actual kernel, will involve more complex setup, but the VM is essentially unaware of these differences.

Now that this groundwork is laid, continuing on to point 2, here are my (admittedly minimally informed) observations:

* We have a kernel, the transaction kernel, which orchestrates the rollup execution. I think the foundational kernel executable described above provides everything needed by the transaction kernel, so no additional abstraction is needed.
* We have an executable, the transaction script, which targets the transaction kernel. It orchestrates the execution of one or more note scripts, sort of like "shelling out" to an external program on a typical system. Nominally, note scripts are executables, but rather than the kernel being responsible for their execution, the transaction script is. This seems to call for some kind of abstraction for process management in the kernel, i.e. the transaction script needs to be able to "spawn" a note script for execution, inheriting the context of the transaction script itself (or perhaps with a more restricted context, similar to how this can be done on a typical Linux system). Thus, a transaction script is really two things:
  - An executable program that executes a transaction (i.e. the "script" in transaction script)
  - A container for a set of executable programs that are orchestrated by the script
* Lastly, we have account code, which is effectively just library code. Not sure there is any additional abstraction needed here - this code is simply referenced from one or more of the artifact types mentioned above.

Looking at all this collectively, to me it seems a lot more clear (to me anyway) where our current model of things is lacking, and where additional definition/clarity (and some refactoring probably) would set us up for a much clearer path moving forward.

Do you have any thoughts on things I've missed above, or different ways of categorizing/structuring things that you think might be better than what I've laid out above? I think the biggest open questions in my mind are:

1. What (if any) issues/limitations arise from putting kernels in control of executing a program (say, a transaction script). In what ways is that a departure from how we've been doing things (or planning on doing them) so far?
2. What would process management (or something along those lines) look like in a kernel implementation? How would that tie in, or be affected by, any capability model/permissions/auth we have in place or want to implement in the future?
3. How do these various elements get represented in the package format, and how much does the VM need to know about any of this in order to provide all the functionality we need? I _think_ it's the case that the VM can be ignorant of anything outside of category 1 primitives, and possibly even the distinction between kernel and non-kernel executables, but that last one might depend a little bit on how we interpret packages at the VM level.